cleo_plugins/Audio/Audio.cpp

```diff
@@ -0,0 +1,320 @@
+        CLEO_RegisterOpcode(0x2503, opcode_2503); // set_audio_stream_speed
+        CLEO_RegisterOpcode(0x2504, opcode_2504); // set_audio_stream_volume_with_transition
+        CLEO_RegisterOpcode(0x2505, opcode_2505); // set_audio_stream_speed_with_transition
+
+        // register event callbacks
+        CLEO_RegisterCallback(eCallbackId::GameBegin, OnGameBegin);
+        CLEO_RegisterCallback(eCallbackId::GameProcess, OnGameProcess);
+        CLEO_RegisterCallback(eCallbackId::GameEnd, OnGameEnd);
+        CLEO_RegisterCallback(eCallbackId::DrawingFinished, OnDrawingFinished);
+        CLEO_RegisterCallback(eCallbackId::MainWindowFocus, OnMainWindowFocus);
+    }
+
+    static void __stdcall OnGameBegin(DWORD saveSlot)
+    {
+        soundSystem.Init();
+    }
+
+    static void __stdcall OnGameProcess()
+    {
+        soundSystem.Update();
+    }
+
+    static void __stdcall OnGameEnd()
+    {
+        soundSystem.UnloadAllStreams();
+    }
+
+    static void __stdcall OnDrawingFinished()
+    {
+        if (CTimer::m_UserPause) // main menu visible
+            soundSystem.Update();
+    }
+
+    static void __stdcall OnMainWindowFocus(bool active)
+    {
+        soundSystem.Update(active);
+    }
+
+ 
+    //0AAC=2,  %2d% = load_audiostream %1d%  // IF and SET
+    static OpcodeResult __stdcall opcode_0AAC(CScriptThread* thread)
+    {
+        auto path = OPCODE_READ_PARAM_FILEPATH();
+
+        auto ptr = soundSystem.LoadStream(path);
+
+        OPCODE_WRITE_PARAM_PTR(ptr);
+        OPCODE_CONDITION_RESULT(ptr != nullptr);
+        return OR_CONTINUE;
+    }
+
+    //0AAD=2,set_audiostream %1d% perform_action %2d%
+    static OpcodeResult __stdcall opcode_0AAD(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM(); VALIDATE_STREAM()
+        auto action = OPCODE_READ_PARAM_INT();
+
+        if (stream)
+        {
+            switch (action)
+            {
+                case 0: stream->Stop(); break;
+                case 1: stream->Play(); break;
+                case 2: stream->Pause(); break;
+                case 3: stream->Resume(); break;
+                default:
+                    LOG_WARNING(thread, "Unknown audiostream's action (%d) in script %s", action, ScriptInfoStr(thread).c_str());
+            }
+        }
+
+        return OR_CONTINUE;
+    }
+
+    //0AAE=1,release_audiostream %1d%
+    static OpcodeResult __stdcall opcode_0AAE(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+
+        soundSystem.UnloadStream(stream);
+
+        return OR_CONTINUE;
+    }
+
+    //0AAF=2,%2d% = get_audiostream_length %1d%
+    static OpcodeResult __stdcall opcode_0AAF(CScriptThread* thread)
+    {
+        auto stream = (CAudioStream*)OPCODE_READ_PARAM_PTR(); VALIDATE_STREAM();
+
+        auto length = stream->GetLength();
+        length /= max(stream->GetSpeed(), 0.000001f); // speed corrected
```

Comment:
is this to avoid division by 0?

Reply:
Yes

Reply:
I feel like there should be a way to get unscaled stream length. If you watch a video on 2x speed the player does not change the total length. Maybe this opcode should return length as-is, and new 2501 get_audiostream_length_scaled would return it scaled to the speed?

Reply:
To get unscaled length just do not change the default speed (1.0). It is only affected by user set speed, not by game speed multiplier.
General idea was to keep it consistent with already existing behaviour as previously there was no way to change speed. It seems to make sense to ask "how long will this track play". Then use it as wait opcode argument.

Reply:
> It seems to make sense to ask "how long will this track play". 

this what new 2501 opcode is for. I don't see why we need two opcodes doing same

Reply:
Original opcode was returning length as seconds in form of floored integer.
Updated to keep the old opcode behaviour unchanged.

Reply:
Ok. Last thing is the 2501 name. "Fine" is ambiguous word and can mean different things. It's probably unclear what this command does without reading the description.

Let's call it GET_AUDIO_STREAM_LENGTH_SCALED_TO_SPEED. A little verbose, but has clear meaning.

Alternatives: GET_AUDIO_STREAM_LENGTH_SCALED
GET_AUDIO_STREAM_DURATION